zebrad/src/commands/seed.rs

```diff
@@ -0,0 +1,155 @@
+//! `seed` subcommand - runs a dns seeder
+
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use abscissa_core::{config, Command, FrameworkError, Options, Runnable};
+use futures::channel::oneshot;
+use tower::{buffer::Buffer, Service, ServiceExt};
+use zebra_network::{AddressBook, BoxedStdError, Request, Response};
+
+use crate::{config::ZebradConfig, prelude::*};
+
+/// Whether our `SeedService` is poll_ready or not.
+#[derive(Debug)]
+enum SeederState {
+    /// Waiting for the address book to be shared with us via the oneshot channel.
+    AwaitingAddressBook(oneshot::Receiver<Arc<Mutex<AddressBook>>>),
+    /// Address book received, ready to service requests.
+    Ready(Arc<Mutex<AddressBook>>),
+}
+
+#[derive(Debug)]
+struct SeedService {
+    state: SeederState,
+}
+
+impl Service<Request> for SeedService {
+    type Response = Response;
+    type Error = BoxedStdError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        debug!("SeedService.state: {:?}", self.state);
+
+        match self.state {
+            SeederState::Ready(_) => return Poll::Ready(Ok(())),
+            SeederState::AwaitingAddressBook(ref mut rx) => match rx.try_recv() {
+                Err(e) => {
+                    error!("SeedService oneshot sender dropped: {:?}", e);
+                    return Poll::Ready(Err(e.into()));
+                }
+                Ok(None) => {
+                    trace!("SeedService hasn't received a message via the oneshot yet.");
+                    return Poll::Pending;
+                }
+                Ok(Some(address_book)) => {
+                    info!(
+                        "SeedService received address_book via oneshot {:?}",
+                        address_book
+                    );
+                    self.state = SeederState::Ready(address_book);
+                    return Poll::Ready(Ok(()));
+                }
+            },
+        }
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        info!("SeedService handling a request: {:?}", req);
+
+        let address_book = if let SeederState::Ready(address_book) = &self.state {
+            address_book
+        } else {
+            panic!("SeedService::call without SeedService::poll_ready");
+        };
+
+        let response = match req {
+            Request::GetPeers => {
+                debug!(address_book.len = address_book.lock().unwrap().len());
+                info!("SeedService responding to GetPeers");
+                Ok::<Response, Self::Error>(Response::Peers(
+                    address_book.lock().unwrap().peers().collect(),
+                ))
+            }
+            _ => Ok::<Response, Self::Error>(Response::Ok),
```

Comment:
subsequent to this PR, we should do the plumbing required to turn errors from the internal service into `reject` messages, and change this from an `Ok` to an `Err`.

Reply:
https://github.com/ZcashFoundation/zebra/issues/94